packages/react-composites/src/composites/CallWithChatComposite/adapter/AzureCommunicationCallWithChatAdapter.ts

```diff
@@ -499,6 +500,97 @@ export const createAzureCommunicationCallWithChatAdapter = async ({
+  /**
+   * Arguments to be passed to {@link createAzureCommunicationCallWithChatAdapter}.
+   *
+   * Allows arguments to be undefined so that you can respect the rule-of-hooks and pass in arguments
+   * as they are created. The adapter is only created when all arguments are defined.
+   */
+  args: Partial<AzureCommunicationCallWithChatAdapterArgs>,
+  /**
+   * Optional callback to modify the adapter once it is created.
+   *
+   * If set, must return the modified adapter.
+   */
+  afterCreate?: (adapter: CallWithChatAdapter) => Promise<CallWithChatAdapter>,
+  /**
+   * Optional callback called before the adapter is disposed.
+   *
+   * This is useful for clean up tasks, e.g., leaving any ongoing calls.
+   */
+  beforeDispose?: (adapter: CallWithChatAdapter) => Promise<void>
+): CallWithChatAdapter | undefined => {
+  const { credential, displayName, endpoint, locator, userId } = args;
+  const [adapter, setAdapter] = useState<CallWithChatAdapter | undefined>(undefined);
+  const afterCreateRef = useRef(afterCreate);
+  const beforeDisposeRef = useRef(beforeDispose);
+
+  // These refs are updated on *each* render, so that the latest values
+  // are used in the `useEffect` closures below.
+  // Using a Ref ensures that new values for the callbacks do not trigger the
+  // useEffect blocks, and a new adapter creation / distruction is not triggered.
+  afterCreateRef.current = afterCreate;
+  beforeDisposeRef.current = beforeDispose;
+
+  useEffect(
+    () => {
+      if (!credential || !displayName || !endpoint || !locator || !userId) {
+        return;
+      }
+      (async () => {
+        let newAdapter = await createAzureCommunicationCallWithChatAdapter({
+          credential,
+          displayName,
+          endpoint,
+          locator,
+          userId
+        });
+        if (afterCreateRef.current) {
+          newAdapter = await afterCreateRef.current(newAdapter);
+        }
+        setAdapter(newAdapter);
+      })();
+    },
+    // Explicitly list all arguments so that caller doesn't have to memoize the `args` object.
+    [afterCreateRef, credential, displayName, endpoint, locator, userId]
+  );
+
+  // Dispose the old adapter when a new one is created.
+  // This clean up function is returned from a separate `useEffect` block because:
+  // - `adapter` is set asynchronously, so the clean up for `useEffect` that creates the adapter can not capture
+  //   the value of `adapter` that needs to be disposed.
+  // - `adapter` can not be added to the dependency array of `useEffect` above -- we do not want to trigger a new
+  //   adapter creation because of the first adapter creation.
+  useEffect(() => {
+    return () => {
+      (async () => {
+        if (adapter) {
+          if (beforeDisposeRef.current) {
+            await beforeDisposeRef.current(adapter);
+          }
+          adapter.dispose();
+        }
+      })();
+    };
+  }, [adapter, beforeDisposeRef]);
```

Comment:
Is it OK that we construct, assing, use a new adapter before disposing the old one?
i.e.
- adapter is created
- displayName updated
- new adapter created
  - afterCreateCalled()
- ... some async time passes 
- oldAdapterDispose async function is triggered
- ... some async time passes 
- oldAdapter is disposed

Reply:
Should we call dispose in front of `let newAdapter = await createAzureCommunicationCallWithChatAdapter({` also?

Reply:
Great catch. You are right, we aren't allowed to have to adapters (actually two `CallAgent` instances) at the same time.

I added a dumb rename loop in the sample app to trigger this behavior: 
![image](https://user-images.githubusercontent.com/82062616/157300840-875965f3-a8c1-4325-9484-fb25a2831889.png)

and did see failure in creating adapter.

The new commit fixes this issue. We need another `useRef` so that we don't cause infinite adapter creation loop.

Look at my very stupid sample app that keeps renaming the local participant :D 

https://user-images.githubusercontent.com/82062616/157300969-7ffb15e6-7d71-43ab-85a4-97fc1a8207ae.mp4

Reply:
Of course, if you have a new adapter, you reset the `CallWithChatComposite`, so that you end up back on the configuration page.

Reply:
Hehe love the video!